### DISCUSSION TITLE
Using Shairport-Sync together with mkchromecast and ALSA

### DISCUSSION THREAD

**Discussion by MrMeeb** [mikebrady/shairport-sync#1727] [Q&A]
Hey,

I've got a couple Sonos Play:1s kicking around - they're Sonos speakers that predate Sonos' support for AirPlay 2. The app is slow and cludgy, so I was interested in getting them to work with AirPlay 2. I was keen to go with AP2 specifically, because it supports multiple speakers, giving me multi-room.

In comes mkchromecast (https://github.com/muammar/mkchromecast). It supports sending a Linux device's system audio to either a chromecast or sonos device. It by default does this via PulseAudio, but also supports ALSA. The ALSA method (https://github.com/muammar/mkchromecast/wiki/ALSA) seems to be to create some kind of loobpack device that captures all the desktop audio, routing it back into mkchromecast.

My idea is that combining Shairport-Sync and mkchromecast should result in AirPlay 2-compatible older Sonos devices. 

Problem is, I can't get mkchromecast to work with ALSA. It works fine out of the box with PulseAudio, but following the instructions for ALSA leaves me with 0 audio coming out of the system speakers or Sonos device. Similarly, I can get Shairport-Sync working no issues with ALSA, coming out of the system speakers, but couldn't really get it going with PulseAudio. Since Shairport-Sync's recommendation is ALSA, I wanted to go that route anyway. 

So it seems like I can get either half of the puzzle sort of working, but can't combine it all together. The main roadblock I think being I can't seem to get mkchromecast working with ALSA. 

Since mkchromecast's developer has stated he isn't really working on the project anymore, I'm appealing over here to see if anyone would be kind enough to help out. I'm reasonably comfortable with using Linux as a server platform, but working with its various audio tools is completely new to me since servers aren't usually playing music!

Been testing on Ubuntu Desktop 22.04 LTS. In the case of Shairport-Sync, I downloaded and built it manually instead of running it in Docker. I am comfortable with Docker, just thought it'd be faster and remove a layer of abstraction for initial testing.

Any help would be hugely appreciated.

**Comment by MrMeeb** [comment]
Something I just found by going down the route of PulseAudio - it works, but there's a delay of about 8-10s between the device receiving audio from Shairport-Sync and it coming out of the Sonos. Since SS has no control for sync in the audio pipeline beyond itself (as I understand), I guess this becomes quite fragile when doing multiple speakers select in AP2. It makes video completely unuseable, but my main goal was music anyway. Even so, when using multiple airplay2 endpoints, each attached to individual speakers, the chance of them being out of sync seems high.

Apparently when using ALSA with mkchromecast, the delay drops to 'maximum 4 seconds', so maybe less room for error, although my desired multi-room solution might not work quite as well as I'd hoped if everything beyond Shairepoint-Sync is out of sync

**Comment by mikebrady** [comment]
Thanks for the interesting discussion. TBH I don't know too much about Sonos or mkchromecast. I would agree that the whole setup would be fragile. If there was code to drive those old Sonos speakers directly and get latency information back from them, that would be the way to go.

**Reply by MrMeeb** [reply]
Thanks Mike. Mkchromecast uses SoCo, but I believe at the base level it creates a webserver that constantly streams a sound file. which it then instructs the Sonos devices to stream from. So you aren't really controlling the hardware, just interfacing with it the same way you would with an app. 

I don't have the fundamental understanding of ALSA or really coding of this kind that's required to create an application that captures data from shairport-sync and ouputs it to Sonos, especially since it's far from clear whether the syncing would work at the end. So I think the journey stops here for me.

Cheers for your input.

**Reply by MrMeeb** [reply]
Follow up!

I ditched Mkchromecast and instead found a method to send ALSA audio via a loopback device to Sonos using an adaptation of the method described here: https://gist.github.com/denysvitali/6d15e0cab8c4851d76ca19356d30be05

Audio is output to the loopback device via shairport-sync. ffmpeg captures this audio using the ALSA loopback device as an input, transcodes to AAC, and pipes it to VLC. VLC then exposes it over http as an mp3 file. You can then instruct Sonos to play this file. I do so using soco-cli.

There are still some unknowns, and I'd like to bundle it all into a docker container; currently it's a VM per speaker. I'd also like to be able to control volume on the Sonos devices using the volume hook that Shairport-sync supports. However this is an interesting POC. There is a delay of ~5-7 seconds between hitting play on the device and it playing on the Sonos devices. Acceptable for just audio playback. I cannot notice any difference in sync between the two Sonos devices, at least over a short playback session, which is really encouraging. Even a minor playback sync difference wouldn't be a massive issue.
